integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java

```diff
@@ -296,21 +353,27 @@ private void launchTachyonWorkerContainers(List<Container> containers) {
         FormatUtils.getSizeFromBytes((long) mRamdiskMemInMB * Constants.MB));
 
     for (Container container : containers) {
-      if (mAllWorkersAllocatedLatch.getCount() == 0) {
-        break;
-      }
-      try {
-        // Launch container by create ContainerLaunchContext
-        ContainerLaunchContext ctx = Records.newRecord(ContainerLaunchContext.class);
-        ctx.setCommands(commands);
-        ctx.setEnvironment(environmentMap);
-        LOG.info("Launching container {} for Tachyon worker {} on {} with worker command: {}",
-            container.getId(), mNumWorkers - mAllWorkersAllocatedLatch.getCount(),
-            container.getNodeHttpAddress(), command);
-        mNMClient.startContainer(container, ctx);
-        mAllWorkersAllocatedLatch.countDown();
-      } catch (Exception ex) {
-        LOG.error("Error launching container " + container.getId() + " " + ex);
+      synchronized (mWorkerHosts) {
+        if (mWorkerHosts.size() >= mNumWorkers
+            || mWorkerHosts.contains(container.getNodeId().getHost())) {
+          LOG.info("Releasing assigned container on {}", container.getNodeId().getHost());
+          // Avoid re-using nodes - we don't support multiple workers on the same node
+          mRMClient.releaseAssignedContainer(container.getId());
+        } else {
+          try {
+            ContainerLaunchContext ctx = Records.newRecord(ContainerLaunchContext.class);
+            ctx.setCommands(commands);
+            ctx.setEnvironment(environmentMap);
+            LOG.info("Launching container {} for Tachyon worker {} on {} with worker command: {}",
+                container.getId(), mWorkerHosts.size(), container.getNodeHttpAddress(), command);
+            mNMClient.startContainer(container, ctx);
+            mWorkerHosts.add(container.getNodeId().getHost());
+          } catch (Exception ex) {
```

Comment:
I just changed the indentation of your code :P

Switched to new style throughout file.